[Feat] Full Text Search 이용한 게시물 검색 기능 구현

jeong1112 opened the pull request:
## Related issue 🛠
- closed #149
  
## Work Description 📝
- Full Text Search를 이용하여 검색 기능을 구현하였습니다.
- Full Text Search 전용 인덱스를 생성해야 하는데, 기획 정책 상 `제목 + 내용 + 작성자`로 검색이 가능하기 때문에 인덱스를 생성하려면 Post 테이블에 작성자 컬럼이 존재해야 해서 생성했습니다.
- 컬럼이 새로 생성됨에 따라, 유저 닉네임 수정 시 게시물 작성자도 동기화할 수 있도록 로직을 추가했습니다.
## Uncompleted Tasks 😅
Ngram Parser가 토큰을 파싱하는 기본 사이즈가 2여서 현재는 검색어가 2자 이상일 때만 검색 기능이 작동하는데, 검색어가 한 글자인 경우에 대해서, 이를 like를 통해 검색하는 것이 나을지, `ngram_token_size`를 1로 설정할지 고민입니다. 사이즈를 1로 설정하게 되면 문자 하나하나가 별개의 토큰이 되어서 검색 결과의 정확성이 떨어진다고 하는데, 한 글자인 경우 그냥 Like를 통한 검색을 해도 된다고 보시나요?

jeong1112 committed:
[Feat] #149 Post 테이블에 작성자명 컬럼 추가

jeong1112 committed:
[Feat] #149 닉네임 수정 시 게시물 작성자명도 동기화되도록 구현

jeong1112 committed:
[Feat] #149 Full Text Search function 생성 메서드 구현

jeong1112 committed:
[Feat] #149 Full Text Search 로직 구현

jeong1112 committed:
[Feat] #149 게시물 검색 기능 구현

jeong1112 committed:
[Chore] #149 Full Text Search 메서드 네이밍 수정

jher235 commented on src/main/java/org/sopt/bofit/domain/post/controller/PostController.java:
```diff
@@ -173,4 +173,16 @@ public BaseResponse<Void> createCommentReply(
+            @RequestParam(required = false, defaultValue = POSTS_DEFAULT_SIZE) int size
+    ){
+        return BaseResponse.ok(postService.searchPosts(keyword, cursorId, size),"게시물 검색 성공");
+    }
```
저희 지난 회의 때 조회 api에서 전부 처리하는 형태로 이야기 하지 않았나요??
물론 현재 형태가 나쁘다고 생각하진 않습니다~

jher235 commented on src/main/java/org/sopt/bofit/domain/post/repository/PostCustomRepositoryImpl.java:
```diff
@@ -97,5 +99,48 @@ public Slice<PostSummaryResponse> findAllByCursorId(Long cursorId, int size) {
+        return new SliceImpl<>(content, PageRequest.of(0, size), hasNext);
+    }
+
+    private NumberExpression<Double> getRelevanceScore(String keyword, QPost post) {
+        return Expressions.numberTemplate(Double.class,
+                "function('match_language_mode',{0},{1},{2},{3})",
+                post.title, post.content, post.writerNickname, keyword
+        );
+    }
```
단순 네이티브 쿼리로 처리하실 줄 알았는데 이렇게 alias로 처리할 수 있는 좋은 방법이 있네요~~ 👍

jher235 commented on src/main/java/org/sopt/bofit/domain/post/repository/PostCustomRepositoryImpl.java:
```diff
@@ -97,5 +99,48 @@ public Slice<PostSummaryResponse> findAllByCursorId(Long cursorId, int size) {
+        if (hasNext) content.remove(size);
+
+        return new SliceImpl<>(content, PageRequest.of(0, size), hasNext);
+    }
```
사용자의 좋아요 여부 + 좋아요 수도 함께 보여야 하지 않나요?
그리고 comment 쪽을 count 로 처리하는게 DB 부하가 너무 심할 것 같아서... 개선이 필요해보입니다..

jeong1112 replied:
그래서 저희가 전에 얘기한 것처럼 commentCount를 컬럼으로 적용하려고 했는데 그렇게 하는 게 좋을까요?

jher235 replied:
확인이 늦었네요.. 죄송합니다
지금은 기능적으로 문제없이 동작하니까 괜찮지만 이번 스프린트 feature 를 다 쳐낸 뒤에 수정하면 좋을 것 같습니다!!

jher235 reviewed (approved):
고생하셨습니다~~~
개인적으로는 최소 2-gram 이 좋다고는 생각하는데,, 기획과 상의가 필요해보이네요 
인덱스는 따로 DB 단에서 거실 예정이신가요?

jeong1112 committed:
[Feat] #149 Post에 댓글 수 컬럼으로 추가

jeong1112 committed:
[Feat] #149 댓글 생성 / 삭제 시 commentCount 반영

# Conflicts:
#	src/main/java/org/sopt/bofit/domain/comment/service/CommentService.java

jeong1112 committed:
[Chore] #149 게시물 조회 시 댓글과 조인하지 않도록 수정

jeong1112 committed:
[Chore] #149 게시물 조회 DTO들에 좋아요 수, 좋아요 여부 추가

jeong1112 committed:
[Chore] #149 게시물 조회 로직에 좋아요 수, 좋아요 여부 추가

jeong1112 commented:
@jher235 commentCount 컬럼 추가, 좋아요 여부 및 개수 DTO 추가한 사항입니다!

jeong1112 commented:
++ 인덱스는 DB 단에 걸어놨습니다
